## 3.4
### Correlation matrix:
$Rx = 
\begin{bmatrix}
1 & 0.5\\
0.5 & 1\\
\end{bmatrix}
$
### Condition for the learning rate to be convergent: 


$0 < \eta < \frac{1}{lambda_{\mathrm{max}}}\\ or \\0 < \eta < \frac{1}{trace_{\mathrm{Rx}}}$

### Calculate eigenvalues and trace:

In [2]:
import numpy as np
m = [[1, 0.5], [0.5, 1]]
eigenv = np.linalg.eig([[1, 0.5], [0.5, 1]])
lambda_max = max(eigenv[0])

trace = np.trace(m)

print("e1 = ", eigenv[0][0],"\n"
      "e2 = ", eigenv[0][1],"\n"
      "trace = ", trace,"\n")
 

e1 =  1.5 
e2 =  0.5 
trace =  2.0 




### Result:
Upperbound (eigenvalues): $\frac{1}{lambda_{\mathrm{max}}} = \frac{1}{1.5} = \frac{2}{3}$   
Upperbound (trace): $\frac{1}{lambda_{\mathrm{max}}} = \frac{1}{2}$  
LMS is converging for (using eigenvalues) : $0 < \eta < \frac{2}{3}$  
LMS is converging for (using trace) : $0 < \eta < \frac{1}{2}$ 